internal/database/txn/transaction.go

```diff
@@ -170,7 +170,47 @@ func (t *RetryingTxnRunner) Txn(ctx context.Context, db *sqlair.DB, fn func(cont
 		}
 
 		if err := fn(ctx, tx); err != nil {
-			if rErr := t.retryStrategy(ctx, tx.Rollback); rErr != nil {
+			if rErr := t.retryStrategy(ctx, func(context.Context) error {
+				return tx.Rollback()
+			}); rErr != nil {
+				t.logger.Warningf("failed to rollback transaction: %v", rErr)
+			}
+			return errors.Trace(err)
+		}
+
+		return errors.Trace(t.commit(ctx, tx))
+	})
+}
```

Comment:
`Txn(...)` should probably just become:
```
func (t *RetryingTxnRunner) Txn(ctx context.Context, db *sqlair.DB, fn func(context.Context, *sqlair.TX) error) error {
    return t.TxnWithPrecheck(ctx, db, noopPrecheck, fn)
}
func noopPrecheck(ctx context.Context) error {
    select {
        case <-ctx.Done():
            return ctx.Err():
        default:
            return nil
    }
}
```

or

```
func (t *RetryingTxnRunner) Txn(ctx context.Context, db *sqlair.DB, fn func(context.Context, *sqlair.TX) error) error {
    return t.TxnWithPrecheck(ctx, db, noopPrecheck, fn)
}
func noopPrecheck(_ context.Context) error {
    return nil
}
```

Reply:
and StdTxn will one day be toast, so it doesn't matter here.

Reply:
I tried to remove the StdTxn in this PR, but it turned out to be too much work. I did keep the work though https://github.com/juju/juju/pull/18144